apm-opentracing/src/main/java/co/elastic/apm/opentracing/ExternalProcessSpanContext.java

```diff
@@ -30,16 +30,36 @@
 import java.util.Map;
 
 class ExternalProcessSpanContext implements ApmSpanContext {
+
+    // co.elastic.apm.agent.impl.transaction.TraceContext - only for the purpose of holding the parsed IDs
+    @Nullable
+    private Object childTraceContext;
```

Comment:
Just an idea: instead of eagerly allocating the TraceContext object, could we instead validate the header and just parse out the trace id and span id on demand?

Reply:
Thought about all kind of stuff in order to reduce the allocations here, but I felt we can live with the overhead in this case on the favor of simplification.

Reply:
Yeah, we definitely can. But we don't have to.

More often than not, the `toTraceId`/`toSpanId` methods won't be called (especially `ExternalProcessSpanContext` and allocating a TraceContext just in case seems a bit wasteful.

It wouldn't be any more complex I think. IMHO it's ok to just allocate a TraceContext each time `toTraceId`/`toSpanId` is called (it even says on the javadoc that it may cause allocations). So we don't even need a field for it.

We could then later further optimize it by not allocating a TraceContext object but to parse just the trace id/span id out of the header, without any changes in the bridge classes.

I don't insist on that though - I just think it would be a nice and simple improvement.

Reply:
Ahh, I misunderstood you. I thought you meant adding code to parse the ID on demand from the raw text map. I will just use this field and set it lazily on demand only if null, then we get one allocation max.

Reply:
> I thought you meant adding code to parse the ID on demand from the raw text map.

That's actually what I meant 😄

Reply:
Well, I would really like to avoid duplicating the traceparent header parsing code for this.
I now only parse and create a trace context lazily if the getId APIs are used, and only once.
I believe that addresses your allocation concern, right?